pkg/http/sender.go

```diff
@@ -0,0 +1,638 @@
+
+	// Authentication-related fields
+	user    string
+	pass    string
+	token   string
+	tlsMode tlsMode
+
+	client http.Client
+}
+
+// LineSenderOption defines line sender option.
+type LineSenderOption func(s *LineSender)
+
+// WithTls enables TLS connection encryption.
+func WithTls() LineSenderOption {
+	return func(s *LineSender) {
+		s.tlsMode = tlsEnabled
+	}
+}
+
+// WithBasicAuth sets a Basic authentication header for
+// ILP requests over HTTP.
+func WithBasicAuth(user, pass string) LineSenderOption {
+	return func(s *LineSender) {
+		s.user = user
+		s.pass = pass
+	}
+}
+
+// WithBearerToken sets a Bearer token Authentication header for
+// ILP requests.
+func WithBearerToken(token string) LineSenderOption {
+	return func(s *LineSender) {
+		s.token = token
+	}
+}
+
+// WithRequestTimeout is used in combination with min_throughput
+// to set the timeout of an ILP request. Defaults to 10 seconds.
+//
+// timeout = (request.len() / min_throughput) + request_timeout
+func WithRequestTimeout(timeout time.Duration) LineSenderOption {
+	return func(s *LineSender) {
+		s.requestTimeout = timeout
+	}
+}
+
+// WithMinThroughput is used in combination with request_timeout
+// to set the timeout of an ILP request. Defaults to 100KiB/s.
+//
+// timeout = (request.len() / min_throughput) + request_timeout
+func WithMinThroughput(bytesPerSecond int) LineSenderOption {
+	return func(s *LineSender) {
+		s.minThroughputBytesPerSecond = bytesPerSecond
+	}
+}
+
+// WithRetryTimeout is the cumulative maximum duration spend in
+// retries. Defaults to 10 seconds.
+//
+// Only network-related errors and certain 5xx response
+// codes are retryable.
+func WithRetryTimeout(t time.Duration) LineSenderOption {
+	return func(s *LineSender) {
+		s.retryTimeout = t
+	}
+}
+
+// WithInitBuffer size sets the desired initial buffer capacity
+// in bytes to be used when sending ILP messages. Defaults to 0.
+func WithInitBufferSize(sizeInBytes int) LineSenderOption {
+	return func(s *LineSender) {
+		s.InitBufSizeBytes = sizeInBytes
+	}
+}
+
+// WithAddress sets address to connect to. Should be in the
+// "host:port" format. Defaults to "127.0.0.1:9000".
+func WithAddress(addr string) LineSenderOption {
+	return func(s *LineSender) {
+		s.address = addr
+	}
+}
+
+// WithTlsInsecureSkipVerify enables TLS connection encryption,
+// but skips server certificate verification. Useful in test
+// environments with self-signed certificates. Do not use in
+// production environments.
+func WithTlsInsecureSkipVerify() LineSenderOption {
+	return func(s *LineSender) {
+		s.tlsMode = tlsInsecureSkipVerify
+	}
+}
+
+// WithAutoFlushDisabled turns off auto-flushing behavior.
+// To send ILP messages, the user must call Flush().
+func WithAutoFlushDisabled() LineSenderOption {
+	return func(s *LineSender) {
+		s.autoFlush = false
+	}
+}
+
+// WithAutoFlushRows sets the number of buffered rows that
+// must be breached in order to trigger an auto-flush.
+// Defaults to 75000.
```

Comment:
Looks like in Java this setting defaults to 600. Worth double-checking other clients and having a common default.

Reply:
> Looks like in Java this setting defaults to 600. Worth double-checking other clients and having a common default.

@amunra the auto-flush rows default in the ILP Client Configuration document is 75k. Should I use 75k or 600 here?